rpcs3/Emu/Io/GunCon3.cpp

```diff
@@ -0,0 +1,301 @@
+			 + data[4] + data[5] ^ data[6] ^ data[7])
+			 + data[8] + data[9] - data[10] - data[11] ^ data[12];
+}
+
+usb_device_guncon3::usb_device_guncon3(u32 controller_index, const std::array<u8, 7>& location)
+	: usb_device_emulated(location)
+	, m_controller_index(controller_index)
+{
+	device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE,
+		UsbDeviceDescriptor {
+			.bcdUSB             = 0x0110,
+			.bDeviceClass       = 0x00,
+			.bDeviceSubClass    = 0x00,
+			.bDeviceProtocol    = 0x00,
+			.bMaxPacketSize0    = 0x08,
+			.idVendor           = 0x0b9a,
+			.idProduct          = 0x0800,
+			.bcdDevice          = 0x8000,
+			.iManufacturer      = 0x00,
+			.iProduct           = 0x00,
+			.iSerialNumber      = 0x00,
+			.bNumConfigurations = 0x01});
+	auto& config0 = device.add_node(UsbDescriptorNode(USB_DESCRIPTOR_CONFIG,
+		UsbDeviceConfiguration {
+			.wTotalLength        = 0x0020,
+			.bNumInterfaces      = 0x01,
+			.bConfigurationValue = 0x01,
+			.iConfiguration      = 0x00,
+			.bmAttributes        = 0x00,
+			.bMaxPower           = 0x32}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_INTERFACE,
+		UsbDeviceInterface {
+			.bInterfaceNumber   = 0x00,
+			.bAlternateSetting  = 0x00,
+			.bNumEndpoints      = 0x02,
+			.bInterfaceClass    = 0xff,
+			.bInterfaceSubClass = 0x00,
+			.bInterfaceProtocol = 0x00,
+			.iInterface         = 0x00}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT,
+		UsbDeviceEndpoint {
+			.bEndpointAddress = 0x02,
+			.bmAttributes     = 0x03,
+			.wMaxPacketSize   = 0x0008,
+			.bInterval        = 0x10}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT,
+		UsbDeviceEndpoint {
+			.bEndpointAddress = 0x82,
+			.bmAttributes     = 0x03,
+			.wMaxPacketSize   = 0x000f,
+			.bInterval        = 0x04}));
+}
+
+usb_device_guncon3::~usb_device_guncon3()
+{
+}
+
+void usb_device_guncon3::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue, u16 wIndex, u16 wLength, u32 buf_size, u8* buf, UsbTransfer* transfer)
+{
+	transfer->fake            = true;
+	transfer->expected_count  = buf_size;
+	transfer->expected_result = HC_CC_NOERR;
+	transfer->expected_time   = get_timestamp() + 100;
+
+	usb_device_emulated::control_transfer(bmRequestType, bRequest, wValue, wIndex, wLength, buf_size, buf, transfer);
+}
+
+extern bool is_input_allowed();
+
+void usb_device_guncon3::interrupt_transfer(u32 buf_size, u8* buf, u32 endpoint, UsbTransfer* transfer)
+{
+	transfer->fake = true;
+	transfer->expected_count = buf_size;
+	transfer->expected_result = HC_CC_NOERR;
+	transfer->expected_time = get_timestamp() + 500;
+
+	if ((endpoint & 0x80) == LIBUSB_ENDPOINT_OUT)
+	{
+		std::memcpy(m_key, buf, std::min(buf_size, static_cast<u32>(sizeof(m_key))));
+		return;
+	}
+	// else ENDPOINT_IN
+
+	ensure(buf_size >= sizeof(struct GunCon3_data));
+	std::memset(buf, 0, buf_size);
+	struct GunCon3_data* gc = reinterpret_cast<struct GunCon3_data*>(buf);
```

Comment:
```suggestion
	GunCon3_data gc{};

	// ... fill gc ...

	std::memcpy(buf, &gc, sizeof(struct GunCon3_data));

	if (const size_t size = buf_size - sizeof(struct GunCon3_data))
	{
		std::memset(buf + buf_size, 0, buf_size - sizeof(struct GunCon3_data));
	}
```

Reply:
You don't need the memset?

Reply:
No, wMaxPacketSize is equal with sizeof(GunCon3_data), so the struct should fill the entire buffer. The games shouldn't ask for more.